### PR TITLE
data-table__filter

### PR DESCRIPTION
**Pull request by xwatkins**
### Reference to issue
Add functionality to enable the filtering of column content.

### Description of changes
Add a data attribute which works as a trigger to add a filter. The content is parsed, and a `<select/>` is added with the corresponding options. When an option is selected, the content of the non-matching rows is hidden with css.

Also dug up the testing config which I'd implemented during last winter's "hackathon". This allows for more testing of the component using the JSDOM and @testing-library functionality.

### Tests / Styleguide
 - [x] Tests pass
 - [x] Styleguide

### CODE REVIEW

File: packages/protvista-datatable/src/protvista-datatable.ts [hunk 1]
```diff
@@ -262,9 +300,42 @@ class ProtvistaDatatable extends LitElement {
...
+      }
+    });
+    return hasMatch;
+  }
+
   updateRowStyling(): void {
     let oddOrEvenCount = 0;
     this.rows?.forEach((row) => {
+      // Filter visibility
+      const isRowVisible = this.isRowVisible(row);
```
Comment (dlrice) [inline packages/protvista-datatable/src/protvista-datatable.ts:333]: Could be a little easier to understand if all of the determination (including L385-395) of a row being visible is within `isRowVisible`

**Review by dlrice — CHANGES_REQUESTED** [review]
I believe there is a bug here where if a row is expanded to reveal its hidden contents and then a filter applied so that its parents disappear, the hidden contents are still visible.

File: packages/protvista-datatable/src/protvista-datatable.ts [hunk 2]
```diff
@@ -262,9 +300,42 @@ class ProtvistaDatatable extends LitElement {
...
+
+  isRowVisible(row: HTMLTableRowElement): boolean {
+    if (!this.selectedFilters || this.selectedFilters.size <= 0) {
+      return true;
+    }
+    let hasMatch = false;
+    this.selectedFilters.forEach((value, filterName) => {
+      const column = row.querySelector(`[data-filter="${filterName}"]`);
+      if (column && column.innerHTML === value) {
+        hasMatch = true;
```
[the remark this answers is not in the document]
Comment (xwatkins) [reply packages/protvista-datatable/src/protvista-datatable.ts:323]: It's not possible to short-circuit `forEach` (unless throwing an exception, which would be overkill). If we didn't have a Map we could use `Array.some`, but as the Map is unlikely to be very big I don't think there would be much of a performance improvement by optimising this.

File: packages/protvista-datatable/src/protvista-datatable.ts (hunk 1 above)
Comment (xwatkins) [reply packages/protvista-datatable/src/protvista-datatable.ts:347]: Moved some of the logic out.

**Comment by xwatkins** [comment]
Comments addressed, and fixed the issue when expanding groups and filtering at the same time.

File: packages/protvista-datatable/src/protvista-datatable.ts [hunk 3]
```diff
@@ -262,9 +300,56 @@ class ProtvistaDatatable extends LitElement {
...
+        column = groupRow.querySelector(`[data-filter="${filterName}"]`);
+      } else {
+        column = row.querySelector(`[data-filter="${filterName}"]`);
+      }
+      if (column && column.innerHTML.replace(/(<([^>]+)>)/gi, "") === value) {
+        hasMatch = true;
+      }
+    });
+    return hasMatch && isExpandedGroup;
+  }
```
[the remark this answers is not in the document]
Comment (xwatkins) [reply packages/protvista-datatable/src/protvista-datatable.ts]: Done.

File: packages/protvista-datatable/src/protvista-datatable.ts [hunk 4]
```diff
@@ -262,9 +300,56 @@ class ProtvistaDatatable extends LitElement {
...
+      if (row.dataset.groupFor) {
+        // If group, get group row
+        const groupRow = this.querySelector(
+          `[data-id="${row.dataset.groupFor}"]`
+        );
+        column = groupRow.querySelector(`[data-filter="${filterName}"]`);
+      } else {
+        column = row.querySelector(`[data-filter="${filterName}"]`);
+      }
+      if (column && column.innerHTML.replace(/(<([^>]+)>)/gi, "") === value) {
```
[the remark this answers is not in the document]
Comment (xwatkins) [reply packages/protvista-datatable/src/protvista-datatable.ts]: Good idea. I think the strip HTML would be more useful in the context of a "text search", which we might want to add later...

**Pull request merged**